### PR TITLE
[#2954] feat(all): add case-sensitive capability

### PR DESCRIPTION
**Pull request by mchades**
### What changes were proposed in this pull request?

- support case-sensitive capability for name
- Hive catalog use case-insensitive capability

### Why are the changes needed?

Fix: #2954 

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

tests added

### CODE REVIEW

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java [hunk 1]
```diff
@@ -25,4 +25,18 @@ public CapabilityResult columnDefaultValue() {
...
+
+  @Override
+  public CapabilityResult caseSensitiveOnName(Scope scope) {
+    switch (scope) {
+      case SCHEMA:
+      case TABLE:
+      case COLUMN:
+        // Hive is case insensitive, see
+        // https://cwiki.apache.org/confluence/display/Hive/User+FAQ#UserFAQ-AreHiveSQLidentifiers(e.g.tablenames,columnnames,etc)casesensitive?
+        return CapabilityResult.unsupported("Hive is case insensitive.");
```
Comment (jerryshao) [inline catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java:37]: Can you please add more logs here?

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java [hunk 2]
```diff
@@ -25,4 +25,18 @@ public CapabilityResult columnDefaultValue() {
         "The DEFAULT constraint for column is only supported since Hive 3.0, "
             + "but the current Gravitino Hive catalog only supports Hive 2.x.");
   }
+
+  @Override
+  public CapabilityResult caseSensitiveOnName(Scope scope) {
```
Comment (jerryshao) [inline catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java:30]: Where did you put the capability restrictions to the catalog scope?

File: core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java [hunk 3]
```diff
@@ -25,22 +33,148 @@ public static TableChange[] applyCapabilities(Capability capabilities, TableChan
...
+              if (change instanceof TableChange.ColumnChange) {
+                return applyCapabilities((TableChange.ColumnChange) change, capabilities);
 
-              } else if (change instanceof TableChange.UpdateColumnNullability) {
-                return applyCapabilities(
-                    (TableChange.UpdateColumnNullability) change, capabilities);
-
-              } else if (change instanceof TableChange.UpdateColumnDefaultValue) {
-                return applyCapabilities(
-                    ((TableChange.UpdateColumnDefaultValue) change), capabilities);
```
Comment (jerryshao) [inline core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java:37]: Why do you remove this?

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java (hunk 1 above)
Comment (mchades) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java:37]: Add more logs or comments?  More to describe what?

File: core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java (hunk 3 above)
Comment (mchades) [reply core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java:37]: Do not remove, just move to method `private static TableChange applyCapabilities(
      TableChange.ColumnChange change, Capability capabilities)`: https://github.com/datastrato/gravitino/pull/2975/files#diff-df7bf73c020a2a9709fe067700e70054ab51dd29b071308e03d618b4607fa5a3R158-R168

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java (hunk 1 above)
Comment (jerryshao) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java:37]: I mean  the unsupported message, like you mentioned in comment.

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java (hunk 2 above)
Comment (mchades) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java:30]: I'm afraid the cability of Catalog scope cannot be managed within the framework. For example, when altering a catalog, we need to first obtain the capability of the catalog, which requires the provider and properties of the catalog. However, we only have the `NameIdentifier` of the catalog. So I plan to remove the catalog scope. Do you have a better solution?

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java (hunk 1 above)
Comment (mchades) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java:37]: There is just one more link in the comment, should I put the link in the message?

Comment (jerryshao) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java:37]: It' fine.

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java (hunk 2 above)
Comment (jerryshao) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java:30]: So how do we restrict the metalake and catalog name?

Comment (mchades) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogCapability.java:30]: added, please help review again

File: catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/IcebergCatalogOperations.java [hunk 4]
```diff
@@ -128,7 +128,7 @@ public NameIdentifier[] listSchemas(Namespace namespace) throws NoSuchCatalogExc
           icebergTableOps.listNamespace(IcebergTableOpsHelper.getIcebergNamespace()).namespaces();
 
       return namespaces.stream()
-          .map(icebergNamespace -> NameIdentifier.of(icebergNamespace.levels()))
+          .map(icebergNamespace -> NameIdentifier.of(namespace, icebergNamespace.toString()))
```
Comment (mchades) [inline catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/IcebergCatalogOperations.java:131]: @FANNG1 Could you please help review this change, thx

File: core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java [hunk 5]
```diff
@@ -25,22 +35,212 @@ public static TableChange[] applyCapabilities(Capability capabilities, TableChan
...
+
+  public static Index[] applyCapabilities(Index[] indexes, Capability capabilities) {
+    return Arrays.stream(indexes)
+        .map(i -> applyCapabilities(i, capabilities))
+        .toArray(Index[]::new);
+  }
+
+  public static NameIdentifier applyCapabilitiesOnMetalake(NameIdentifier ident) {
+    String sensitiveOnMetalakeName = applyCaseSensitiveOnMetalakeName(ident.name());
+    applyNameSpecificationOnMetalake(sensitiveOnMetalakeName);
```
Comment (jerryshao) [inline core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java:125]: Maybe `NameSpec` is enough, the method name seems too long?

File: core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java [hunk 6]
```diff
@@ -25,22 +35,212 @@ public static TableChange[] applyCapabilities(Capability capabilities, TableChan
...
+    Namespace namespace = ident.namespace();
+
+    String sensitiveOnMetalakeName = applyCaseSensitiveOnMetalakeName(namespace.level(0));
+    namespace.levels()[0] = sensitiveOnMetalakeName;
+    applyNameSpecificationOnMetalake(namespace.level(0));
+
+    String sensitiveOnCatalogName = applyCaseSensitiveOnCatalogName(ident.name());
+    applyNameSpecificationOnCatalog(ident.name());
+    return NameIdentifier.of(namespace, sensitiveOnCatalogName);
+  }
```
Comment (jerryshao) [inline core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java:140]: I would suggest remove the metalake and catalog related check in this PR. I think it is out of the scope of the Capability framework, we can do it separately.

File: core/src/main/java/com/datastrato/gravitino/catalog/FilesetOperationDispatcher.java [hunk 7]
```diff
@@ -43,9 +45,20 @@ public FilesetOperationDispatcher(
    */
   @Override
   public NameIdentifier[] listFilesets(Namespace namespace) throws NoSuchSchemaException {
+    NameIdentifier standardizedSchemaIdent =
+        standardizeNameIdentifier(NameIdentifier.of(namespace.levels()), Capability.Scope.SCHEMA);
+    Namespace standardizedNamespace =
+        Namespace.of(
+            standardizedSchemaIdent.namespace().level(0),
+            standardizedSchemaIdent.namespace().level(1),
+            standardizedSchemaIdent.name());
```
Comment (jerryshao) [inline core/src/main/java/com/datastrato/gravitino/catalog/FilesetOperationDispatcher.java:54]: Can we have a wrapper method like `doWithCatalog` to standardize the input namespace or nameidentifier, like:

```java
doWithStandardizedIdent { ident ->
   doWithXXX
 }
```

Comment (jerryshao) [reply core/src/main/java/com/datastrato/gravitino/catalog/FilesetOperationDispatcher.java:54]: The current change placed everywhere makes code hard to review.

File: catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/IcebergCatalogOperations.java (hunk 4 above)
Comment (FANNG1) [reply catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/IcebergCatalogOperations.java:131]: I think it's ok for current fix

File: core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java [hunk 8]
```diff
@@ -25,22 +35,212 @@ public static TableChange[] applyCapabilities(Capability capabilities, TableChan
...
+        namespace.levels()[namespace.length() - 1] = sensitiveOnSchemaName;
+        applyNameSpecification(SCHEMA, sensitiveOnSchemaName, capabilities);
+        break;
+      default:
+        throw new IllegalArgumentException("Unexpected capability scope: " + scope);
+    }
+    String name = applyCaseSensitiveOnName(scope, ident.name(), capabilities);
+    applyNameSpecification(scope, name, capabilities);
+    return NameIdentifier.of(namespace, name);
+  }
```
[the remark this answers is not in the document]
Comment (mchades) [reply core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java:75]: refactored

File: core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java (hunk 6 above)
Comment (mchades) [reply core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java]: removed, please help review again, thx

File: core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java [hunk 9]
```diff
@@ -25,22 +33,165 @@ public static TableChange[] applyCapabilities(Capability capabilities, TableChan
...
+    return Arrays.stream(expressions)
+        .map(e -> applyCapabilities(e, capabilities))
+        .toArray(Expression[]::new);
+  }
+
+  private static Expression applyCapabilities(Expression expression, Capability capabilities) {
+    if (expression instanceof NamedReference.FieldReference) {
+      NamedReference.FieldReference ref = (NamedReference.FieldReference) expression;
+      String[] fieldName = applyCaseSensitiveOnColumnName(ref.fieldName(), capabilities);
+      ref.fieldName()[0] = fieldName[0];
```
Comment (jerryshao) [inline core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java:147]: Also here, I think it is not good to do in place modification, do you have better solution?

File: core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java [hunk 10]
```diff
@@ -25,22 +33,165 @@ public static TableChange[] applyCapabilities(Capability capabilities, TableChan
...
+
+    } else if (change instanceof TableChange.UpdateColumnDefaultValue) {
+      return applyCapabilities(((TableChange.UpdateColumnDefaultValue) change), capabilities);
+
+    } else if (change instanceof TableChange.RenameColumn) {
+      return applyCapabilities((TableChange.RenameColumn) change, capabilities);
+
+    } else {
+      String[] fieldName = applyCaseSensitiveOnColumnName(change.fieldName(), capabilities);
+      change.fieldName()[0] = fieldName[0];
```
Comment (jerryshao) [inline core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java:189]: Also here.

File: core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java [hunk 11]
```diff
@@ -89,6 +240,20 @@ private static TableChange applyCapabilities(
         updateColumnDefaultValue.getNewDefaultValue());
   }
 
+  private static TableChange applyCapabilities(
+      TableChange.RenameColumn renameColumn, Capability capabilities) {
+    String[] fieldName = applyCaseSensitiveOnColumnName(renameColumn.fieldName(), capabilities);
+    String newName = renameColumn.getNewName();
+    if (fieldName.length == 1) {
```
Comment (jerryshao) [inline core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java:247]: What if length >=1?

File: core/src/main/java/com/datastrato/gravitino/catalog/FilesetOperationDispatcher.java (hunk 7 above)
Comment (jerryshao) [reply core/src/main/java/com/datastrato/gravitino/catalog/FilesetOperationDispatcher.java]: I think you didn't change this, do you have a good solution?

File: core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java (hunk 11 above)
Comment (mchades) [reply core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java:247]: length >=1 means renaming nested field, not column, so no need to handle

File: core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java [hunk 12]
```diff
@@ -89,6 +240,20 @@ private static TableChange applyCapabilities(
...
+      applyNameSpecification(Capability.Scope.COLUMN, fieldName[0], capabilities);
+      newName =
+          applyCaseSensitiveOnName(
+              Capability.Scope.COLUMN, renameColumn.getNewName(), capabilities);
+      applyNameSpecification(Capability.Scope.COLUMN, newName, capabilities);
+    }
+    return TableChange.renameColumn(fieldName, newName);
+  }
+
```
[the remark this answers is not in the document]
Comment (mchades) [reply core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java:256]: The target object `XXX` has been reflected in the parameters, so I think `XXX` in the method name can be omitted.

BTW, when the basic capability `A` of the application is used, I use method name `applyA` to express it, such as `applyColumnNotNull`, `applyColumnDefaultValue`

**Comment by jerryshao** [comment]
@mchades One thing I think we can improve is to use wrapper design pattern https://zhuanlan.zhihu.com/p/43368677.

We can have one dispatcher that handles all the capability things, another dispatcher that handles operation logic, one another dispatcher that does the event posting. using wrapper pattern to chain them altogether like using `InputStream`. With this, all the related logic will only reside in the related class, no need to change the existing a lot. which makes the code easy to understand, can you please think a bit on this?

**Comment by mchades** [comment]
> For example, like adding new `CapabilityAppliedFilesetDispatcher`, and chain with `FilesetEventDispatcher` then `FilesetOperationDispatcher`. Like:
> 
> ```java
>   public NameIdentifier[] listFilesets(Namespace namespace) throws NoSuchSchemaException {
>     try {
>       // Apply the capabilities to fileset
>       // Call `FilesetEventDispatcher`'s listFilesets
>       NameIdentifier[] nameIdentifiers = dispatcher.listFilesets(namespace);
>       return nameIdentifiers;
>     } catch (Exception e) {
>       eventBus.dispatchEvent(
>           new ListFilesetFailureEvent(PrincipalUtils.getCurrentUserName(), namespace, e));
>       throw e;
>     }
>   }
> ```

I'm doing the refactor, but I would name the class `xxxStandardizedDispatcher`(such as `TableStandardizedDispatcher`) and it focus on standardize the input and out of the `xxxOperations`, because some other capabilities, such as `managedStorage` need to use internally in `xxxOperations`. WDYT? @jerryshao

**Comment by jerryshao** [comment]
OK, I think we can do it as possible as we can.

**Comment by mchades** [comment]
> OK, I think we can do it as possible as we can. But maybe not just standardize the name, others like check nullable, default value, it can also be done in this dispatcher?

yes

File: core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java (hunk 9 above)
Comment (mchades) [reply core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java]: refactored

**Comment by mchades** [comment]
@jerryshao Please help review again when you have time, thx

File: core/src/main/java/com/datastrato/gravitino/catalog/FilesetStandardizedDispatcher.java [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+import com.datastrato.gravitino.Namespace;
+import com.datastrato.gravitino.connector.capability.Capability;
+import com.datastrato.gravitino.exceptions.FilesetAlreadyExistsException;
+import com.datastrato.gravitino.exceptions.NoSuchFilesetException;
+import com.datastrato.gravitino.exceptions.NoSuchSchemaException;
+import com.datastrato.gravitino.file.Fileset;
+import com.datastrato.gravitino.file.FilesetChange;
+import java.util.Map;
+
+public class FilesetStandardizedDispatcher implements FilesetDispatcher {
```
Comment (jerryshao) [inline core/src/main/java/com/datastrato/gravitino/catalog/FilesetStandardizedDispatcher.java:19]: How about `FilesetNormalizeDispatcher`, change to `xxxNormalizeDispatcher`?

File: core/src/main/java/com/datastrato/gravitino/catalog/FilesetStandardizedDispatcher.java [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        applyCapabilities(ident, Capability.Scope.FILESET, capability),
+        applyCapabilities(capability, changes));
+  }
+
+  @Override
+  public boolean dropFileset(NameIdentifier ident) {
+    return dispatcher.dropFileset(standardizeNameIdentifier(ident));
+  }
+
+  private NameIdentifier standardizeNameIdentifier(NameIdentifier ident) {
```
Comment (jerryshao) [inline core/src/main/java/com/datastrato/gravitino/catalog/FilesetStandardizedDispatcher.java:72]: `normalizeNameIdentifier`?

File: core/src/main/java/com/datastrato/gravitino/catalog/FilesetStandardizedDispatcher.java (hunk 13 above)
Comment (mchades) [reply core/src/main/java/com/datastrato/gravitino/catalog/FilesetStandardizedDispatcher.java]: updated

**Comment by jerryshao** [comment]
@mchades there're some style issues, please fix them.

**Review by jerryshao — APPROVED** [review]

**Pull request merged**